### ISSUE TITLE
[JIT] Simple dispatch overhead benchmark is 4x+ slower than python

### ISSUE THREAD

**Issue by jamesr66a** [pytorch/pytorch#55283]
```
import torch, time

NITER = 1000000
PROCESSOR_HZ = 2e9

x = torch.randn(1)
y = torch.randn(1)

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

tps = (e - s) / NITER
print('single element add benchmark')
print('time per iter', tps)
print('cycles per iter', tps * PROCESSOR_HZ)
print('')
"""
single element add benchmark
time per iter 1.8831143379211427e-06
cycles per iter 3766.228675842285
"""

@torch.jit.script
def bench(x, y, NITER : int):
    rv = torch.zeros([1])
    for _ in range(NITER):
        rv = x + y
    return rv
s = time.time()
bench(x, y, NITER)
e = time.time()

tps = (e - s) / NITER
print('single element add benchmark (torchscript)')
print('time per iter', tps)
print('cycles per iter', tps * PROCESSOR_HZ)
print('')
"""
single element add benchmark (torchscript)
time per iter 8.86375117301941e-06
cycles per iter 17727.50234603882
"""

s = time.time()
for _ in range(NITER):
    pass
e = time.time()

tps = (e - s) / NITER
print('empty python loop baseline')
print('time per iter', tps)
print('cycles per iter', tps * PROCESSOR_HZ)
"""
empty python loop baseline
time per iter 2.9577016830444336e-08
cycles per iter 59.15403366088867
"""
```

Perf profile with call-graph indicates there's some string construction happening in `torch::jit::is_enabled`:

![image](https://user-images.githubusercontent.com/4685384/113495813-cd04ea00-94a8-11eb-97df-ed18c8801c87.png)

https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/csrc/jit/jit_log.cpp#L55

Which ultimately comes from shape profiling:

https://github.com/pytorch/pytorch/blob/041b4431b28eb24335189979e9a110dc9892dfe4/torch/csrc/jit/runtime/profiling_record.cpp#L157

cc @gmagogsfm

**Comment by jamesr66a** [comment]
Gutting `torch::jit::is_enabled`:

```
diff --git a/torch/csrc/jit/jit_log.cpp b/torch/csrc/jit/jit_log.cpp
index d08dff1822d..2437e6e9ec2 100644
--- a/torch/csrc/jit/jit_log.cpp
+++ b/torch/csrc/jit/jit_log.cpp
@@ -53,6 +53,7 @@ static std::unordered_map<std::string, size_t> parseJITLogOption(
 }
 
 bool is_enabled(const char* cfname, JitLoggingLevels level) {
+  return false;
   static const char* c_log_level = std::getenv("PYTORCH_JIT_LOG_LEVEL");
   static const std::unordered_map<std::string, size_t> files_to_levels =
       parseJITLogOption(c_log_level);
```

Knocks it down to 11k cycles:

```
single element add benchmark
time per iter 1.9091026782989504e-06
cycles per iter 3818.205356597901

single element add benchmark (torchscript)
time per iter 5.580108404159546e-06
cycles per iter 11160.216808319092

empty python loop baseline
time per iter 2.891421318054199e-08
cycles per iter 57.828426361083984
```

A lot of overhead from shape profiling seems to remain:
![image](https://user-images.githubusercontent.com/4685384/113496534-73ec8480-94af-11eb-89ae-761b56615cd4.png)
![image](https://user-images.githubusercontent.com/4685384/113496538-7e0e8300-94af-11eb-8a6f-66277cbbe6bb.png)
![image](https://user-images.githubusercontent.com/4685384/113496544-8961ae80-94af-11eb-8fc4-d00c908fa3c2.png)
![image](https://user-images.githubusercontent.com/4685384/113496553-9383ad00-94af-11eb-9980-05f6234a7df1.png)

**Comment by jamesr66a** [comment]
Legacy executor performance looks pretty good:

```
import torch, time

NITER = 1000000
PROCESSOR_HZ = 2e9

x = torch.randn(1)
y = torch.randn(1)

doit = True

if doit:
    s = time.time()
    for _ in range(NITER):
        x + y
    e = time.time()

    tps = (e - s) / NITER
    print('single element add benchmark')
    print('time per iter', tps)
    print('cycles per iter', tps * PROCESSOR_HZ)
    print('')
"""
single element add benchmark
time per iter 1.9227466583251953e-06
cycles per iter 3845.4933166503906
"""

torch._C._jit_set_profiling_executor(False)

@torch.jit.script
def bench(x, y, NITER : int):
    rv = torch.zeros([1])
    for _ in range(NITER):
        #while True:
        rv = x + y
    return rv
s = time.time()
bench(x, y, NITER)
e = time.time()

print(torch.jit.last_executed_optimized_graph())

tps = (e - s) / NITER
print('single element add benchmark (torchscript)')
print('time per iter', tps)
print('cycles per iter', tps * PROCESSOR_HZ)
print('')
"""
single element add benchmark (torchscript)
time per iter 1.2950563430786132e-07
cycles per iter 259.01126861572266
"""

s = time.time()
for _ in range(NITER):
    pass
e = time.time()

tps = (e - s) / NITER
print('empty python loop baseline')
print('time per iter', tps)
print('cycles per iter', tps * PROCESSOR_HZ)
"""
empty python loop baseline
time per iter 2.942395210266113e-08
cycles per iter 58.847904205322266
"""
```

**Comment by gmagogsfm** [comment]
Took a quick look and it seems the most overhead comes from profiling cost in the first runs.

On my machine, TorchScript single element add benchmark takes:

91k cycles per iteration without warm-up
2.6k cycles per iteration after 2 iterations of warm-up
2.57k cycles per iteration with legacy executor

As a comparison, same computation takes 21k cycles per iteration in eager.

The difference between profiling executor and legacy executor seems small enough to be justifiable as noise. So my gut feeling is that the 4x slow down almost all come from the first few profiling runs, the fact that this benchmark is small magnified the fixed overhead in profiling executor.

Next useful thing to look at is overhead in those few profiling runs, one thing we can do is changing code to force profiling run to always trigger and use `perf` to analyze bottleneck.

**Comment by eellison** [comment]
According to @gmagogsfm benchmarking, this isn't an issue except for first two runs. Closing, feel free to reopen if you think there is more to discuss.
